File: client/src/reducers/game.js
const initialState = {
  id: null,
  inProgress: true,
  winner: null,
  currentPlayer: 'player1',
  lastPlayer: null,
  lastTurn: false
}

export default (state = initialState, action) => {
  switch (action.type) {
    case 'NEW_GAME':
      const id = action.gameData.id || state.id

      return { ...state, id }

    case 'CHANGE_CURRENT_PLAYER':
      const currentPlayer = state.currentPlayer === 'player1'
        ? 'player2'
        : 'player1'

      return { ...state, currentPlayer: currentPlayer }
      
    case 'DISPLAY_GAME':
      return action.game

    case 'LAST_TURN':
      const lastPlayer = state.currentPlayer === 'player1' ? 'player2' : 'player1'

      return { ...state, lastTurn: true, lastPlayer }
      
    case 'FINISH_GAME':
      return {
        ...state,
        winner: action.winner,
        inProgress: false
      }

    default:
      return state
  }
}

File: app/models/game.rb
class Game < ApplicationRecord
  belongs_to :user
  has_many :scorecards, dependent: :destroy
  
  scope :for, -> (id) { where(user_id: id).order(id: :desc) }

  def self.create_with_scorecards(user, players)
    Game.create(user_id: user.id).tap do |game|
      game.scorecards = Scorecard.create_for(players, game)
      game.save
    end
  end

  def change_player(params, scorecard)
    next_player = params[:player] == 'player1' ? 'player2' : 'player1'

    self.update(current_player: next_player,
                          last_turn: scorecard.score >= 10000)
  end

  def unwin
    self.current_player = self.winner
    self.winner = nil
    self.in_progress = true
    self.last_turn = false
    index = self.current_player.match(/\d/)[0].to_i - 1
    sc = self.scorecards[index]

    sc.score -= sc.scores_per_turn.pop
    sc.save
    self.save
  end

  def removeLastTurn(playerNumber)
    index = playerNumber - 1
    sc = self.scorecards[index]

    sc.score -= sc.scores_per_turn.pop
    sc.save
  end
end
